ethereum/deploy_util.go

```diff
@@ -0,0 +1,133 @@
+package contracts
+
+import (
+	"context"
+	"crypto/ecdsa"
+	"fmt"
+
+	"github.com/athanorlabs/go-relayer/impls/gsnforwarder"
+	"github.com/ethereum/go-ethereum/accounts/abi/bind"
+	ethcommon "github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/ethclient"
+	logging "github.com/ipfs/go-log"
+
+	"github.com/athanorlabs/atomic-swap/ethereum/block"
+)
+
+var log = logging.Logger("ethereum")
+
+// DeploySwapFactoryWithKey deploys the SwapFactory contract using the passed privKey to
+// pay for the gas.
+func DeploySwapFactoryWithKey(
+	ctx context.Context,
+	ec *ethclient.Client,
+	privkey *ecdsa.PrivateKey,
+	forwarderAddress ethcommon.Address,
+) (ethcommon.Address, *SwapFactory, error) {
+
+	txOpts, err := newTXOpts(ctx, ec, privkey)
+	if err != nil {
+		return ethcommon.Address{}, nil, err
+	}
+
+	if (forwarderAddress != ethcommon.Address{}) {
+		// ensure domain separator is registered
+		forwarder, err := gsnforwarder.NewForwarder(forwarderAddress, ec) //nolint:govet // shadow declaration of err
+		if err != nil {
+			return ethcommon.Address{}, nil, err
+		}
+
+		err = registerDomainSeparator(ctx, ec, privkey, forwarderAddress, forwarder)
+		if err != nil {
+			return ethcommon.Address{}, nil, err
+		}
+	}
+
+	address, tx, sf, err := DeploySwapFactory(txOpts, ec, forwarderAddress)
+	if err != nil {
+		return ethcommon.Address{}, nil, fmt.Errorf("failed to deploy swap factory: %w", err)
+	}
+
+	_, err = block.WaitForReceipt(ctx, ec, tx.Hash())
+	if err != nil {
+		return ethcommon.Address{}, nil, err
+	}
+
+	log.Infof("deployed SwapFactory.sol: address=%s tx hash=%s", address, tx.Hash())
+
+	return address, sf, nil
+}
+
+// DeployGSNForwarderWithKey deploys and registers the GSN forwarder using the passed
+// private key to pay the gas fees.
+func DeployGSNForwarderWithKey(
+	ctx context.Context,
+	ec *ethclient.Client,
+	privkey *ecdsa.PrivateKey,
+) (ethcommon.Address, error) {
+
+	txOpts, err := newTXOpts(ctx, ec, privkey)
+	if err != nil {
+		return ethcommon.Address{}, err
+	}
+
+	address, tx, contract, err := gsnforwarder.DeployForwarder(txOpts, ec)
+	if err != nil {
+		return ethcommon.Address{}, fmt.Errorf("failed to deploy Forwarder.sol: %w", err)
+	}
+
+	_, err = block.WaitForReceipt(ctx, ec, tx.Hash())
+	if err != nil {
+		return ethcommon.Address{}, err
+	}
+
+	err = registerDomainSeparator(ctx, ec, privkey, address, contract)
+	if err != nil {
+		return ethcommon.Address{}, err
+	}
+
+	return address, nil
+}
+
+func registerDomainSeparator(
+	ctx context.Context,
+	ec *ethclient.Client,
+	privkey *ecdsa.PrivateKey,
+	address ethcommon.Address,
+	contract *gsnforwarder.Forwarder,
+) error {
+
+	txOpts, err := newTXOpts(ctx, ec, privkey)
+	if err != nil {
+		return err
+	}
+
+	tx, err := contract.RegisterDomainSeparator(txOpts, gsnforwarder.DefaultName, gsnforwarder.DefaultVersion)
```

Comment:
now that I'm thinking about this, it should only register the domain separator if it isn't already set. do you want to add that or should I? the domain separators are stored in the forwarder contract map `domains` and is the hash of a few things, done in `registerDomainSeparator`

Reply:
I'll create a separate PR for that tomorrow.